### PR TITLE
M3-4614 Add: JS Client methods and types for DbaaS

### PR DESCRIPTION
**Pull request by Jskobos**
## Description

JS Client stuff for DbaaS.

Some things are _not_ here, because either the API doesn't support them or we're not likely to work on them for a while and I'm not sure the API is stable, namely:

- Methods for backups
- Method for resizing

### CODE REVIEW

File: packages/api-v4/src/databases/databases.schema.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      'W12',
+      'W14',
+      'W16',
+      'W18',
+      'W20',
+      'W22'
+    ])
+});
+
+export const createDatabaseSchema = object({
```
Comment (Jskobos) [inline packages/api-v4/src/databases/databases.schema.ts:33]: These are fairly simple for now.

File: packages/api-v4/src/vlans/vlans.ts [hunk 2]
```diff
@@ -55,7 +55,7 @@ export const createVlan = (data: CreateVLANPayload) =>
  * Delete a single VLAN
  */
 export const deleteVlan = (vlanID: number) =>
-  Request<VLAN>(
+  Request<{}>(
```
Comment (Jskobos) [inline packages/api-v4/src/vlans/vlans.ts:58]: Oops!

File: packages/api-v4/src/databases/databases.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  );
+
+/**
+ * getDatabase
+ *
+ * Return detailed information about a single database
+ *
+ */
+export const getDatabase = (databaseID: number) =>
+  Request<Page<Database>>(
```
Comment (johnwcallahan) [inline packages/api-v4/src/databases/databases.ts]: Don't think this should be a `Page` here.

File: packages/api-v4/src/databases/types.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  label: string;
+  price: {
+    hourly: number;
+    monthly: number;
+  };
+  memory: number;
+  disk: number;
+  transfer: number | null;
+  vcpus: number;
+  availability: 'standard' | 'high';
```
Comment (johnwcallahan) [inline packages/api-v4/src/databases/types.ts:12]: Should this use the `DatabaseAvailability` type defined below?

File: packages/api-v4/src/databases/databases.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+/**
+ * resetDatabasePassword
+ *
+ * Resets the root password for a database
+ */
+export const resetPassword = (databaseID: number, root_password: string) =>
+  Request<{}>(
+    setURL(`${API_ROOT}/databases/mysql/instances/${databaseID}/password`),
+    setMethod('PUT'),
```
Comment (johnwcallahan) [inline packages/api-v4/src/databases/databases.ts:124]: Huh, I'm surprised this is a PUT instead of POST like Linode Disks.

**Review by johnwcallahan — APPROVED** [review]

File: packages/api-v4/src/databases/types.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  availability: DatabaseAvailability;
+}
+
+export type DatabaseAvailability = 'standard' | 'high';
+
+export type DatabaseStatus = 'initializing' | 'ready' | 'error' | 'unknown';
+
+export type DatabaseBackupStatus =
+  | 'running'
+  | 'started'
```
Comment (DevDW) [inline packages/api-v4/src/databases/types.ts:21]: Based on that .yaml spec doc, should this be `starting`?

File: packages/api-v4/src/databases/types.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  maintenance_schedule: DatabaseMaintenanceSchedule;
+  vcpus: number;
+  memory: number;
+  disk: number;
+  availability: DatabaseAvailability;
+  updated: string;
+  created: string;
+}
+
+export interface CreateDatabasePayload {
```
Comment (DevDW) [inline packages/api-v4/src/databases/types.ts:76]: Should `region` be in here too, and should `label` be required?

**Review by DevDW — COMMENTED** [review]
Left a couple of questions for `types.ts`; depending on the answers, it looks like a couple of tweaks in `databases.schema.ts` might have to be made

File: packages/api-v4/src/databases/types.ts (hunk 6 above)
Comment (Jskobos) [reply packages/api-v4/src/databases/types.ts:21]: Yes, good catch

File: packages/api-v4/src/databases/types.ts (hunk 7 above)
Comment (Jskobos) [reply packages/api-v4/src/databases/types.ts:76]: yes 👍  and no 👎 :

```
❯ curld databases/mysql/instances POST '{}'| jq
{
  "errors": [
    {
      "reason": "region is required",
      "field": "region"
    },
    {
      "reason": "type is required",
      "field": "type"
    },
    {
      "reason": "root_password is required",
      "field": "root_password"
    }
  ]
}
```

**Review by DevDW — APPROVED** [review]

**Pull request merged**